Frag - Partial Rewrite

lambdatiger opened the pull request:
**When merged, this pull request will:**
- Removes old system of tracking rounds by using a per frame event handler and implements [BI projectile event handlers](https://community.bistudio.com/wiki/Arma_3:_Event_Handlers#Projectile_Event_Handlers).
   - Removed spalling requiring multiple object "hitPart" event handlers to be added and removed.
- Separates spall and frag config parameters into separate files.
- More completely defines vanilla ammunition frag parameters.
- Replaces debug functions that used per frame event handlers with those found in https://github.com/acemod/ACE3/pull/9728
   - Small changes to these from the other PR include the removal of reference to `fragCount` config variable.
- Moves repeated config checks and common parameters to separate functions leveraging hashmaps.
- Since blacklisting is local, we can remove the event handlers to avoid having to check projectile variables multiple times.
- Cleaned some magic numbers out of both `doSpall` and `frago` functions.
- Removed unused variables and lines of code with no effects in `frago`.
- Minor changes from https://github.com/acemod/ACE3/pull/9728.
   - Implemented better method for getting list of targets in `frago`.
   - Checking to make sure fragments aren't targeted at playable logic in `frago`.
   - Minor calculation optimizations in `frago` and `doSpall`.
- Fragments and spall now inherit `shotParents` parameters.
   - If someone would like the first parameter (vehicle) of [setShotParents](https://community.bistudio.com/wiki/setShotParents)/[getShotParents](https://community.bistudio.com/wiki/getShotParents) to be preserved, let me know.
- Updated documentation to properly note that config  value `ace_frag_enabled` doesn't do anything.

As always, I appreciate feedback and criticism!

### Todo
- [x] MP testing - small scale (2x player + dedicated)
- [ ] Re-review when more awake
- [ ] String table translations

### IMPORTANT

- [x] If the contribution affects [the documentation](https://github.com/acemod/ACE3/tree/master/docs), please include your changes in this pull request so the documentation will appear on the [website](https://ace3.acemod.org/).
- [x] [Development Guidelines](https://ace3.acemod.org/wiki/development/) are read, understood and applied.
- [x] Title of this PR uses our standard template `Component - Add|Fix|Improve|Change|Make|Remove {changes}`.

johnb432 commented on addons/frag/functions/fnc_addBlacklist.sqf:
```diff
@@ -0,0 +1,29 @@
+params ["_projectile"];
+TRACE_2("addBlackList",_projectile,typeOf projectile);
+
+_projectile setVariable [QGVAR(blacklisted)];
```
Value is missing

lambdatiger replied:
How do I keep doing this?
Fixed here: https://github.com/acemod/ACE3/pull/10157/commits/7b282ec96cfb1130283858f56762c4f2a2e45928

johnb432 commented on addons/frag/functions/fnc_addBlacklist.sqf:
```diff
@@ -0,0 +1,29 @@
+_projectile removeEventHandler [
+    "Explode",
+    _projectile getVariable [QGVAR(explodeEventHandler), -1]
+];
```
```suggestion
];

```

lambdatiger replied:
Fixed here https://github.com/acemod/ACE3/pull/10157/commits/7b282ec96cfb1130283858f56762c4f2a2e45928

johnb432 commented on addons/frag/functions/fnc_doSpall.sqf:
```diff
@@ -1,139 +1,127 @@
+    [_fragment, "purple", true] call FUNC(dev_trackObj);
     #endif
-};
+};
```
```suggestion
};

```

lambdatiger replied:
Fixed here https://github.com/acemod/ACE3/pull/10157/commits/7b282ec96cfb1130283858f56762c4f2a2e45928

lambdatiger commented:
I believe I've found everything I am not blind to after looking at/writing the code. Thanks as always for your help johnb432.

johnb432 commented on addons/frag/functions/fnc_frago.sqf:
```diff
@@ -18,138 +18,107 @@
+    ACE_FRAG_COUNT_MIN_TIME,
+    ACE_FRAG_COUNT_MAX_TIME,
+    (CBA_missionTime - GVAR(lastFragTime)),
+    ACE_FRAG_COUNT_MIN, ACE_FRAG_COUNT_MAX,
```
```suggestion
    ACE_FRAG_COUNT_MIN,
    ACE_FRAG_COUNT_MAX,
```

lambdatiger commented on addons/frag/functions/fnc_fired.sqf:
```diff
@@ -19,40 +19,71 @@
-    TRACE_4("",_localShooter,_unit,_ammo,_projectile);
-    if (!_localShooter) exitWith {};
-    if (_weapon == "Put") exitWith {}; // Ignore explosives placed without ace_explosives
+if (GVAR(reflectionsEnabled) || GVAR(enabled) && _ammo call FUNC(shouldFrag)) then {
```
Aight, sounds good!

PabstMirror commented:
<details>
<summary>All changes in shouldFrag for vanilla ammo</summary>

```
[BombCluster_03_Ammo_F] before:true now:false
[BombCluster_02_Cap_Ammo_F] before:true now:false
[M_Mo_230mm_AT_LG] before:true now:false
[HelicopterExploSmall] before:true now:false
[M_Air_AA_MI02] before:false now:true
[HelicopterExploBig] before:true now:false
[UnderwaterMine_Range_Ammo] before:true now:false
[ammo_Missile_mim145] before:true now:false
[BombCluster_01_Ammo_F] before:true now:false
[G_20mm_HE] before:true now:false
[UnderwaterMineAB_Range_Ammo] before:true now:false
[BombDemine_01_Ammo_F] before:true now:false
[BombCluster_02_UXO_deploy] before:true now:false
[BombDemine_01_SubAmmo_F] before:true now:false
[ACE_SLAMDirectionalMine_Timer_Ammo] before:true now:false
[ClaymoreDirectionalMine_Remote_Ammo_Scripted] before:true now:false
[G_40mm_HEDP_Bullet] before:true now:false
[M_Titan_AA_long] before:false now:true
[SLAMDirectionalMine_Wire_Ammo] before:true now:false
[M_Mo_82mm_AT] before:true now:false
[ammo_Missile_s750] before:true now:false
[BombCluster_01_UXO_deploy] before:true now:false
[G_40mm_HE] before:true now:false
[BombCluster_03_UXO_deploy] before:true now:false
[DemoCharge_Remote_Ammo_Scripted] before:true now:false
[ACE_G_40mm_HE] before:true now:false
[Drone_explosive_ammo_dummy] before:true now:false
[Drone_explosive_ammo] before:true now:false
[R_80mm_HE] before:false now:true
[M_Zephyr_air] before:false now:true
[M_Zephyr_Mi06] before:false now:true
[ClaymoreDirectionalMine_Remote_Ammo] before:true now:false
[M_Zephyr] before:false now:true
[ammo_Missile_LongRangeAABase] before:true now:false
[R_60mm_HE] before:false now:true
[SatchelCharge_Remote_Ammo] before:true now:false
[M_Titan_AA] before:false now:true
[M_Air_AA_MI06] before:false now:true
[M_Mo_230mm_AT] before:true now:false
[ammo_Penetrator_AGM_01] before:true now:false
[SatchelCharge_Remote_Ammo_Scripted] before:true now:false
[M_Mo_82mm_AT_LG] before:true now:false
[DemoCharge_Remote_Ammo] before:true now:false
[Missile_AA_04_F] before:false now:true
[Mo_cluster_AP] before:true now:false
[BombCluster_02_Ammo_F] before:true now:false
[Mo_cluster_AP_UXO_deploy] before:true now:false
[M_Air_AA] before:false now:true
[ACE_SLAMDirectionalMine_Command_Ammo] before:true now:false
[UXO_deploy_BombCluster_base_F] before:true now:false
[UXO_deploy_base_f] before:true now:false
[ammo_Penetrator_AGM_02] before:true now:false
[ACE_G_40mm_HEDP] before:true now:false
[G_40mm_HEDP] before:true now:false
[M_Titan_AA_static] before:false now:true
[Missile_AA_03_F] before:false now:true
[APERSBoundingMine_Range_Ammo] before:true now:false
[UnderwaterMinePDM_Range_Ammo] before:true now:false
```
</details>

`"G_40mm_HE" call ace_frag_fnc_shouldFrag = false` is the big one that I think will be noticed
IMHO we should `force` it so it works like before

PabstMirror commented:
Just a possibility but we could use #10243 's `addExplosionEventHandler`
roughly done in https://github.com/acemod/ACE3/commit/0642d3724dc724ff359b7203469badae9506d312

This will automatically pickup submunitions
Reduce fired eh code as we only have to check should frag on explosions
No race condition on blacklisting

lambdatiger commented:
> All changes in shouldFrag for vanilla ammo
> [long list]
> `"G_40mm_HE" call ace_frag_fnc_shouldFrag = false` is the big one that I think will be noticed IMHO we should `force` it so it works like before

Good catch, I missed forcing the 40mm again. Here's a classification of all of the items mentioned above if you want to provide further feedback on it.

<details>
<summary>Reverted</summary>

```[[G_40mm_HE] before:true now:false
[G_40mm_HEDP_Bullet] before:true now:false
[ACE_G_40mm_HE] before:true now:false
[ACE_G_40mm_HEDP] before:true now:false
[G_20mm_HE] before:true now:false
[ammo_Missile_LongRangeAABase] before:true now:false
```
</details>

<details>
<summary>A number just produce submunitions</summary>

```[BombCluster_03_Ammo_F] before:true now:false 
[BombCluster_02_Cap_Ammo_F] before:true now:false
[BombCluster_01_Ammo_F] before:true now:false
[BombDemine_01_Ammo_F] before:true now:false
[BombCluster_02_UXO_deploy] before:true now:false
[BombCluster_03_UXO_deploy] before:true now:false
[BombCluster_01_UXO_deploy] before:true now:false
[BombCluster_02_Ammo_F] before:true now:false
[Mo_cluster_AP_UXO_deploy] before:true now:false
[UXO_deploy_BombCluster_base_F] before:true now:false (Base class)
[UXO_deploy_base_f] before:true now:false  (Base class)
```
</details>
<details>
<summary>Some would not have created fragments before as previously submunitions did not, but could</summary>

```[Mo_cluster_AP] before:true now:false 
[APERSBoundingMine_Range_Ammo] before:true now:false
[UnderwaterMinePDM_Range_Ammo] before:true now:false
[UnderwaterMine_Range_Ammo] before:true now:false
[UnderwaterMineAB_Range_Ammo] before:true now:false
```
</details>
<details>
<summary>Some shouldn't produce fragments</summary>

```[ammo_Penetrator_AGM_01] before:true now:false -> penetrators shouldn't produce frag, spall is ok
[ammo_Penetrator_AGM_02] before:true now:false -> penetrators shouldn't produce frag, spall is ok
[Drone_explosive_ammo_dummy] before:true now:false -> dummy
```
</details>
<details>
<summary>Some could few fragments in reality and maybe shouldn't be simulated</summary>

```[HelicopterExploSmall] before:true now:false -> this one is up to everyone, but I removed the gas station explosions fragmentation which this inherits from
[HelicopterExploBig] before:true now:false -> this one is up to everyone, but I removed the gas station explosions fragmentation which this inherits from
[ammo_Missile_mim145] before:true now:false -> designed to be a direct contact missile / not many good reasons to sim frag
[ammo_Missile_s750] before:true now:false -> S-350, same reason as MIM-145
[ClaymoreDirectionalMine_Remote_Ammo_Scripted] before:true now:false
[ClaymoreDirectionalMine_Remote_Ammo] before:true now:false
[SatchelCharge_Remote_Ammo] before:true now:false -> Bricks of explosive in a fabric bag
[SatchelCharge_Remote_Ammo_Scripted] before:true now:false
[DemoCharge_Remote_Ammo] before:true now:false -> Bricks of explosive taped together
[BombDemine_01_SubAmmo_F] before:true now:false
[ACE_SLAMDirectionalMine_Timer_Ammo] before:true
[ACE_SLAMDirectionalMine_Command_Ammo] before:true now:false
[SLAMDirectionalMine_Wire_Ammo] before:true now:false
```
</details>
<details>
<summary>Added because they could be</summary>

```[Drone_explosive_ammo] before:true now:false
[M_Air_AA_MI02] before:false now:true
[M_Titan_AA_long] before:false now:true
[DemoCharge_Remote_Ammo_Scripted] before:true now:false
[R_80mm_HE] before:false now:true
[M_Zephyr_air] before:false now:true
[M_Zephyr_Mi06] before:false now:true
[M_Zephyr] before:false now:true
[R_60mm_HE] before:false now:true
[M_Titan_AA] before:false now:true
[M_Air_AA_MI06] before:false now:true
[Missile_AA_04_F] before:false now:true
[M_Air_AA] before:false now:true
```
</details>

lambdatiger commented:
> Just a possibility but we could use #10243 's `addExplosionEventHandler` roughly done in [0642d37](https://github.com/acemod/ACE3/commit/0642d3724dc724ff359b7203469badae9506d312)
> 
> This will automatically pickup submunitions Reduce fired eh code as we only have to check should frag on explosions No race condition on blacklisting

I've elaborated on why I think it may be an issue on your PR, it's why I didn't stick with `init` event handlers in https://github.com/acemod/ACE3/pull/9728. If I'm wrong and it doesn't cause issues, I'd be happy to switch to it.

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+    GVAR(gurney_k) = "1/3";
+};
+
+class APERSBoundingMine_Range_Ammo: MineBase {
```
```suggestion
class BoundingMineBase;
class APERSBoundingMine_Range_Ammo: BoundingMineBase {
```

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+    GVAR(charge) = 3850;
+};
+
+class m_70mm_saami: RocketBase {///!!! fix me
```
```suggestion
class m_70mm_saami: MissileBase {
```

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+    GVAR(skip) = 1;
+};
+
+class R_230mm_fly: RocketBase {
```
```suggestion
class R_230mm_fly: ShellBase {
```

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+    GVAR(classes)[] = {QGVAR(tiny)};
+};
+
+class M_127mm_Firefist_AT: RocketBase { // HOT missile
```
```suggestion
class M_127mm_Firefist_AT: MissileBase { // HOT missile
```

lambdatiger replied:
Completed https://github.com/acemod/ACE3/pull/10157/commits/58e021bda4455b6e3ce59eacb7269764fc7ba8f4

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+class Sh_120mm_APFSDS: ShellBase {
+    GVAR(skip) = 1;
+};
+class Sh_125mm_APFSDS: ShellBase {
```
```suggestion
class Sh_125mm_APFSDS: Sh_120mm_APFSDS {
```

PabstMirror commented on addons/frag/CfgAmmoFragParameters.hpp:
```diff
@@ -0,0 +1,795 @@
+    GVAR(classes)[] = {QGVAR(small), QGVAR(small), QGVAR(medium), QGVAR(large)};
+};
+
+class Sh_125mm_HEAT: Sh_120mm_HE {
```
```suggestion
class Sh_125mm_HEAT: Sh_125mm_HE {
```

PabstMirror replied:
need to shuffle below 744

lambdatiger replied:
Completed at https://github.com/acemod/ACE3/pull/10157/commits/58e021bda4455b6e3ce59eacb7269764fc7ba8f4

PabstMirror commented on addons/frag/functions/fnc_fired.sqf:
```diff
@@ -19,40 +19,71 @@
 
-if (_shouldAdd) then {
-    // firedMan will have nil "_gunner", so just check _unit; for firedVehicle we want to check _gunner
-    private _localShooter = if (isNil "_gunner") then {local _unit} else {local _gunner};
```
we still need some kind of locality check like this
`ace_firedNonPlayer` is going to trigger on all machines for AI

e.g. put down AI and make him throw a grenade and frago will trigger multiple times

lambdatiger replied:
I didn't see that, or at least notice it. I'll have something in an hour here.

lambdatiger replied:
Simplest approach is still best I guess. For reference using `local ((getShotParents _projectile)#1)` is slightly more efficient, but struggles with Zeus controlled units.

lambdatiger commented on addons/frag/functions/fnc_dev_debugAmmo.sqf:
```diff
@@ -17,65 +21,89 @@
-private _allMagsConfigs = configProperties [configFile >> "CfgMagazines", "isClass _x", true];
-private _processedCfgAmmos = [];
+// Gather all configs, either those that could be created from firing or all classes
+private _allAmmoConfigs = [];
```
I updated the function to do what I think it should for the current version of frag and didn't try to hard to be super efficient about it since I was able to run it with RHS/CUP/multiple CDLC loaded in a couple hundred milliseconds. I can definitely do that if I have time tomorrow.

lambdatiger replied:
Did some optimization in https://github.com/acemod/ACE3/pull/10157/commits/185544160af7a5da6300c851c2fe6b34c59c1b55. I don't think there are any other similar searches in the code, and couldn't find any, but please highlight any you see.

lambdatiger commented on addons/frag/XEH_postInit.sqf:
```diff
@@ -1,32 +1,46 @@
-[] call FUNC(dev_startTracing);
+[
+    "CBA_settingsInitialized",
+    {
```
Gotcha, I thought they were ok like that since I it was never brought up in #9728, and the case between the coding guideline's discussion about control flow statement bracket placement and array entries is different. Should be fixed in https://github.com/acemod/ACE3/pull/10157/commits/70606d411cb8fc8fea93e08c9760846615cae216, but feel free to push any other ones you see.

rautamiekka commented on addons/frag/functions/fnc_dev_addRound.sqf:
```diff
@@ -0,0 +1,90 @@
+#include "..\script_component.hpp"
+/*
+ * Author: Lambda.Tiger
+ * This function initialize projectile tracking of a round so that it's path
```
```suggestion
 * This function initialize projectile tracking of a round so that its path
```

rautamiekka commented on addons/frag/functions/fnc_dev_addRound.sqf:
```diff
@@ -0,0 +1,90 @@
+ *
+ * Arguments:
+ * 0: Projectile to be tracked <OBJECT>
+ * 1: Add projectile hit/explode/defelceted event handlers <BOOL> (default: true)
```
```suggestion
 * 1: Add projectile hit/explode/deflected event handlers <BOOL> (default: true)
```

rautamiekka commented on addons/frag/functions/fnc_fired.sqf:
```diff
@@ -19,40 +19,70 @@
-        TRACE_1("Won't frag, but will spall",_caliber);
-        _shouldAdd = true;
-    };
+            // starting v2.18 it may be faster to use the instigator EH parameter, the same as the second entry shotParents, to recreate _shotParent
```
```suggestion
            // as of v2.18 it may be faster to use the instigator EH parameter, the same as the second entry shotParents, to recreate _shotParent
```

lambdatiger replied:
As of implies 2.18 is out (it isn't)

rautamiekka replied:
Had no clue there's a diff.

lambdatiger replied:
Alright, I'm awake now and actually reading it, "as of" would be ok, I was thinking of "as from." Either would work, so I don't think it's worth another commit over it. The next other but I fix I'll reword that comment.

rautamiekka replied:
Fair enough.

lambdatiger commented:
I'll run through and check command cases once HEMTT's winget repository is updated.

lambdatiger commented:
Will publish a branch switching to the common explode event handler when it hits the main branch:
https://github.com/lambdatiger/ace_frag-overhaul/tree/ace-explode-event-handler

lambdatiger commented on addons/frag/functions/fnc_getFragInfo.sqf:
```diff
@@ -0,0 +1,107 @@
+
+params ["_ammo"];
+
+private _ammoInfo = GVAR(fragInfoCache) get _ammo;
+
+if (!isNil "_ammoInfo") exitWith {_ammoInfo};
```
There were disagreements on this approach before, but I can change it again

lambdatiger replied:
done by https://github.com/acemod/ACE3/pull/10157/commits/bdf08541ef3323a4ca746f353f15f130f10a5bef

lambdatiger commented on addons/frag/functions/fnc_getFragInfo.sqf:
```diff
@@ -0,0 +1,107 @@
+if (!isNil "_ammoInfo") exitWith {_ammoInfo};
+
+private _ammoConfig = configFile >> "CfgAmmo" >> _ammo;
+private _fragTypes = [
```
What should be a macro here?

lambdatiger replied:
Ah understood, I thought you wanted me to go full embedded system dev and have everything be macros all the way down.

lambdatiger replied:
done in https://github.com/acemod/ACE3/pull/10157/commits/bdf08541ef3323a4ca746f353f15f130f10a5bef

lambdatiger commented:
#10243 was merged, would it be preferred to create a separate PR or just commit to this PR when migrating to the common explode event?

rautamiekka commented on addons/frag/functions/fnc_getFragInfo.sqf:
```diff
@@ -0,0 +1,99 @@
+    * _geometryCoefficient = 1/2; - cylindrical K factor
+    * _gurneyConstant = 2320; - Gurney constant of tritonal in m/s
+    * Equation - 0.8 for empirical 80% speed
+    * 0.8 * (((_metalMass / _chargeMass) + _geometryCoefficient) ^ - (1 / 2)) * _gurneyConstant;
+    * or 0.8 * _gurneyConstant * sqrt (_chargeMass /(_metalMass + _chargeMass * _geometryCoefficient)); (slightly faster to compute)
```
If the latter is faster, then we shouldn't even give a slower option, knowing ppl.

Hell, although I'm absolutely 1 of the worst to talk math, the latter even looks simpler.
```suggestion
    * 0.8 * _gurneyConstant * sqrt (_chargeMass / (_metalMass + _chargeMass * _geometryCoefficient));
```

lambdatiger replied:
It's honestly mainly to make it clear when you look at the Wikipedia that the same equation is used here and on line 93. The later is an algebraic equivalent, but if you're bad at math or maybe don't take the moment to do the algebra you'd think that line 93 is incorrect simply because it is different.

rautamiekka replied:
I don't doubt the validity (ACE already uses such math in many places I'm happy to have heard of), but if one is faster, then that's the one to use, and in this case it happens, as you said, the latter is faster and also looks simpler. If the other one is needed for clarity like you said, then it should also be included.

lambdatiger commented on addons/frag/functions/fnc_doSpall.sqf:
```diff
@@ -1,139 +1,127 @@
 
-if !(_caliber >= 2.5 || {(_explosive > 0 && {_idh >= 1})}) exitWith {};
-// ACE_player sideChat format ["BBBB"];
+_roundType call FUNC(getSpallInfo) params ["_caliber", "_explosive"];
```
Yeah, that was part of the revert, thanks for the clean-up there.

lambdatiger commented on addons/frag/functions/fnc_frago.sqf:
```diff
@@ -1,155 +1,125 @@
+private _targets = [];
 {
-    {
-        _objects pushBackUnique _x;
-    } forEach (crew _x);
+    private _crew = crew _x;
+    _crew pushBackUnique _x;
+    _targets append _crew;
 } forEach _objects;
```
Did this already, see here https://github.com/acemod/ACE3/pull/9728#discussion_r1454903389

lambdatiger commented on addons/frag/functions/fnc_frago.sqf:
```diff
@@ -1,155 +1,125 @@
-                    (_vecVar / 2) + (random _vecVar),
-                    (_vecVar / 2) + (random _vecVar)
+                    private _vectorDir = _baseVec vectorDiff [
+                        (_vecVar / 2) - (random _vecVar),
+                        (_vecVar / 2) - (random _vecVar),
+                        (_vecVar / 2) - (random _vecVar)
                     ];
```
Look at lines 97-102, I'm not sure what nouber was going for there, but it isn't. We can definitely pre-compute `_vecVar/2`.

lambdatiger replied:
Actually, that's not true, I know what's going on there, I don't know why I didn't just explain it. This code aims around a center point. This is done by this `_vecVar` variable which slightly changes the direction of the fragment, lines 97-102 adjust for the size of the object, and then 106-108 and some randomization to that so it hits over a larger area (still generally just the torso).

lambdatiger replied:
Updated here https://github.com/acemod/ACE3/pull/10157/commits/1568ddef6c81154eb7f3691046b729571a2116bb

lambdatiger commented:
> I think we have a wrapper for setShotParents in common (either event or function), so that should be used instead

The issue isn't that, but that the projectiles are `createVehicleLocal` so while the simulated damage and effects propagate, the projectiles don't and aren't something you have access to on the server (the same way as the master version doSpall and frago). Also there were some issue with setting shot parents, locality and projectiles that aren't networked sync'd like bullets and rockets when I was testing early on (December 2023).
